### DISCUSSION TITLE
Sampling from conditional distribution

### DISCUSSION THREAD

**Discussion by jobstdavid** [kogalur/randomForestSRC#284] [General]
I would like to generate random samples (y_1, ..., y_d) from the bivariate conditional distribution F(y_1,..., y_d | X_1 = x_1, ..., X_p = x_p)? How is this possible? I thought about sampling m samples {(y_1^i, ..., y_d^i)}_{i=1,...m} by using the calculated weights based on (x_1, ..., x_p). This means that we sample m times with replacement from the original sample  {(y_1^i, ..., y_d^i)}_{i=1,...n} according to the random forest weight probabilities. Is this correct or are there alternative/better approaches? Thank you very much in advance!

**Comment by ishwaran** [comment]
It's difficult to sample from a bivariate distribution if it's fully unspecified (i.e. nonparametric simulation).  Might I suggest a different approach?

Estimate the bivariate mean and bivariate variance-covariance matrix and then draw samples from a bivariate normal using the estimated mean and covariance matrix.

Here's how you would estimate the bivariate mean in a toy example using `mtcars` (this returns the bivariate mean for all 32 cases), all estimation using OOB:

```
> o=rfsrc(cbind(mpg,disp)~.,mtcars,forest.wt="oob")
> wt=o$forest.wt
> apply(wt, 1, function(wti) {colSums(o$yvar * wti)})
         [,1]      [,2]      [,3]      [,4]     [,5]      [,6]      [,7]      [,8]     [,9]     [,10]    [,11]     [,12]     [,13]     [,14]    [,15]
mpg   22.1812  21.43223  24.15462  20.55154  15.7523  21.11751  15.38802  22.79478  22.7420  20.28049  20.9339  15.30432  15.41411  15.71849  15.1558
disp 163.1523 172.48838 134.87980 188.75457 324.9347 188.87229 342.61105 156.88929 149.1764 195.01697 190.4931 350.89080 341.62175 336.93783 349.5907
         [,16]     [,17]     [,18]     [,19]     [,20]     [,21]     [,22]     [,23]     [,24]     [,25]     [,26]     [,27]     [,28]     [,29]
mpg   15.28991  14.82982  24.53578  26.52183  26.08049  22.80453  15.61333  16.16563  15.98541  15.43274  26.49437  25.51462  24.02084  17.93981
disp 347.97465 350.92992 125.70481 111.63640 113.24638 159.65590 337.93873 326.49276 324.56933 329.23720 114.96574 124.48734 135.99263 266.52048
         [,30]     [,31]     [,32]
mpg   20.19736  16.46172  24.36478
disp 216.25196 312.94304 133.79137
```

Now to get the covariance matrix, you need to work out the diagonal values and the two off-diagonal values (each of these are the same).

Here's how you would get an estimate for the off-diagonal value of the covariance matrix.  Note this returns 32 values, one for each case:

```
> apply(wt, 1, function(wti) {sum(o$yvar[1] * o$yvar[2] * wti)})
 [1] 3336.968 3429.309 3073.307 3586.824 4963.283 3676.530 5154.053 3296.504 3214.939 3696.287 3705.004 5223.161 5152.110 5130.560 5187.202 5216.250
[17] 5036.039 2932.793 2811.285 2816.355 3354.426 5117.953 5125.199 4942.349 4928.815 2870.902 2910.635 3041.147 4292.992 3901.917 4796.196 3064.905

```

**Reply by jobstdavid** [reply]
Thank you for your response and idea! However, in lots of situations such a parametric assumption is not valid or the best for a certain model/data set. Therefore I would recommend to model the margins individually by the marginal distributions of the response variables which can be easily done with the random forest weights. Afterwards one can join the individual margins via a copula (empirical copula, gauss copula, ...). 
However, I was before just interested if I was right with the sampling scheme based on the forest weights. Have you thought about it?

**Comment by ishwaran** [comment]
I don't believe the sampling scheme you described will work.

Copulas are a better approach and there are many copulas that allow you to match both the marginals and the multivariate correlations.  Therefore, I still think you should estimate the bivariate correlation if you are using that approach.

There may be a third potentially better way to proceed using the OOB residual that I just thought of (this is the idea of Zhang et. al that is used for quantile regression in [`quantreg`](https://www.randomforestsrc.org/reference/quantreg.rfsrc.html)).  The method applies to one-dimension but I believe it can be extended.  Basically the OOB residuals are iid observations and if you are willing to assume a homoskedastic model (which is reasonable), you can simply sample values from the residuals which represent a draw from the error distribution.  Then you just add the predicted value to get your draw.

Here's a simple illustration using mtcars

```
## fit bivariate forest, extract residual
o <- rfsrc(cbind(mpg, disp) ~ ., mtcars)
yhat <-  get.mv.predicted(o)
res <- o$yvar - yhat

## example of a draw for Y
n <- nrow(yhat)
print(yhat +  res[sample(1:n, replace = FALSE),])

                        mpg      disp
Cadillac Fleetwood  17.22039 284.25546
Ford Pantera L      19.02658 263.04481
Pontiac Firebird    27.88189 205.53903
Porsche 914-2       21.42540 179.30551
Ferrari Dino        15.42133 246.16166
Lincoln Continental 15.88928 309.21146
Volvo 142E          12.25532 330.72110
AMC Javelin         22.15268 130.25911
Fiat 128            30.86740 102.16150
Dodge Challenger    20.62542 164.04703
Merc 240D           22.34555 177.06974
Toyota Corona       14.04158 311.62647
Merc 280C           12.04178 328.22240
Hornet 4 Drive      16.84545 405.23393
Merc 450SE          16.47412 270.21476
Camaro Z28          12.10699 388.40934
Merc 230            14.46331 343.16817
Merc 450SLC         24.09389  61.61595
Valiant             24.08623 132.58006
Mazda RX4 Wag       25.43143  99.64925
Datsun 710          21.18965 138.54904
Mazda RX4           14.49326 340.60532
Merc 280            14.82106 306.28742
Lotus Europa        22.08622 280.39244
Merc 450SL          17.41262 258.92424
Honda Civic         30.34359  78.59532
Maserati Bora       23.83284 111.23476
Hornet Sportabout   27.51077 167.08649
Duster 360          17.05329 281.88466
Fiat X1-9           20.85559 187.36323
Toyota Corolla      24.05844 280.26756
Chrysler Imperial   24.54665 219.41281


```
